src/main.rs

```diff
@@ -0,0 +1,94 @@
+#![allow(non_upper_case_globals)]
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+
+include!(concat!(env!("OUT_DIR"), "/cuda_driver_bindings.rs"));
+
+use anyhow;
+use std::ffi::CString;
+use tarpc::{
+    context,
+    server::{BaseChannel, Channel},
+    tokio_serde::formats::Json,
+};
+use futures_util::StreamExt;
+
+mod service;
+use crate::{service::Cuda as CudaService};
+
+// This is the type that implements the generated World trait. It is the business logic
+// and is used to start the server.
+#[derive(Clone)]
+struct RemoteCuda;
+
+#[tarpc::server]
+impl service::Cuda for RemoteCuda {
+    async fn cuInit(self, _: context::Context, flags: u32) -> u32 {
+        // This does not look right. cuInit should be called once per process.
+        // I suspect what we need to do is:
+        // 1. new connection
+        // 2. fork
+        // 3. forked server handles all requests for that connection...
+        unsafe {
+            cuInit(flags)
+        }
+    }
+    async fn cuDeviceGet(self, _: context::Context, ordinal: i32) -> (i32, u32) {
+        let mut dev: CUdevice = 0;
+        let devPtr: *mut CUdevice = &mut dev;
+
+        unsafe {
+            let res = cuDeviceGet(devPtr, ordinal);
+            (*devPtr, res)
+        }
+    }
+    async fn cuDeviceGetCount(self, _: context::Context) -> (i32, u32) {
+        let mut count: i32 = 0;
+        let countPtr: *mut i32 = &mut count;
+
+        unsafe {
+            let res = cuDeviceGetCount(countPtr);
+
+           (*countPtr, res)
+        }
+    }
+    async fn cuDeviceGetName(self, _: context::Context, max_len: i32, dev: i32) -> (String, u32) {
+        let mut name = [0; 1024];
+        let mut len = max_len;
+        if len > 1024 {
+            len = 1024
+        }
+
+        unsafe {
```

Comment:
for the most part, the autogenerated bindings do much of this. I am not sure if bindgen is useful for us.

I now understand though, why you believe autogenerating everything may not be ideal. Tend to agree. For some subset of CUDA functions, auto-generation is a no-go.